File: frontend/src/features/accounts/lib/account-state.ts
import type { AccountStatus } from "@/api";

export type CanonicalAccountState =
  | "healthy"
  | "degraded"
  | "fused"
  | "cost_blocked"
  | "survivor"
  | "paused"
  | "disabled"
  | "excluded"
  | "unknown";

const aliases: Record<string, CanonicalAccountState> = {
  healthy: "healthy",
  active: "healthy",
  available: "healthy",
  normal: "healthy",
  passed: "healthy",
  pass: "healthy",
  succeeded: "healthy",
  success: "healthy",
  通过: "healthy",
  正常: "healthy",
  健康: "healthy",
  degraded: "degraded",
  observing: "degraded",
  降级: "degraded",
  观察中: "degraded",
  fused: "fused",
  fuse_pending: "fused",
  hard_open: "fused",
  soft_open: "fused",
  熔断: "fused",
  已熔断: "fused",
  cost_blocked: "cost_blocked",
  "cost-wall-blocked": "cost_blocked",
  成本墙拦截: "cost_blocked",
  已被成本墙拦截: "cost_blocked",
  survivor: "survivor",
  paused: "paused",
  暂停: "paused",
  已暂停: "paused",
  disabled: "disabled",
  inactive: "disabled",
  停用: "disabled",
  已停用: "disabled",
  excluded: "excluded",
  out_of_scope: "excluded",
  排除: "excluded",
  已排除: "excluded",
};

function normalizeAccountState(value: string | null | undefined): CanonicalAccountState {
  return aliases[(value ?? "").trim().toLowerCase()] ?? "unknown";
}

export function effectiveAccountState(account: AccountStatus): CanonicalAccountState {
  if (account.paused === true) return "paused";
  if (account.health.trim()) return normalizeAccountState(account.health);
  if (account.routing_state?.trim()) return normalizeAccountState(account.routing_state);
  return normalizeAccountState(account.health_status);
}

File: backend/internal/business/account_state.go
package business

import "strings"

const (
	AccountStateHealthy     = "healthy"
	AccountStateDegraded    = "degraded"
	AccountStateFused       = "fused"
	AccountStateCostBlocked = "cost_blocked"
	AccountStateSurvivor    = "survivor"
	AccountStatePaused      = "paused"
	AccountStateDisabled    = "disabled"
	AccountStateExcluded    = "excluded"
	AccountStateUnknown     = "unknown"
)

// NormalizeAccountState is the single backend mapping from upstream and engine
// aliases to the states exposed by account and group read models.
func NormalizeAccountState(value string) string {
	switch strings.ToLower(strings.TrimSpace(value)) {
	case "healthy", "active", "available", "normal", "pass", "passed", "success", "succeeded", "ok", "通过", "正常", "健康":
		return AccountStateHealthy
	case "degraded", "observing", "降级", "观察中":
		return AccountStateDegraded
	case "fused", "fuse_pending", "hard_open", "soft_open", "熔断", "已熔断":
		return AccountStateFused
	case "cost_blocked", "cost-wall-blocked", "成本墙拦截", "已被成本墙拦截":
		return AccountStateCostBlocked
	case "survivor":
		return AccountStateSurvivor
	case "paused", "暂停", "已暂停":
		return AccountStatePaused
	case "disabled", "inactive", "停用", "已停用":
		return AccountStateDisabled
	case "excluded", "out_of_scope", "排除", "已排除":
		return AccountStateExcluded
	default:
		return AccountStateUnknown
	}
}

func accountStatePriority(state string) int {
	switch NormalizeAccountState(state) {
	case AccountStatePaused:
		return 80
	case AccountStateDisabled:
		return 70
	case AccountStateFused:
		return 60
	case AccountStateCostBlocked:
		return 55
	case AccountStateSurvivor:
		return 50
	case AccountStateDegraded:
		return 40
	case AccountStateHealthy:
		return 30
	case AccountStateUnknown:
		return 20
	case AccountStateExcluded:
		return 10
	default:
		return 0
	}
}
